File: Api/src/main/java/net/binggl/mydms/features/senders/SenderStore.java
package net.binggl.mydms.features.senders;

import java.util.List;
import java.util.Optional;

import org.hibernate.Criteria;
import org.hibernate.SessionFactory;
import org.hibernate.criterion.Order;
import org.hibernate.criterion.Restrictions;

import com.google.inject.Inject;

import net.binggl.mydms.features.shared.store.AbstractHibernateStore;

public class SenderStore extends AbstractHibernateStore<Sender> {

	private static final String FIELD_NAME = "name";

	@Inject
	public SenderStore(final SessionFactory factory) {
		super(Sender.class, factory);
	}

	public List<Sender> searchByName(String search) {
		Criteria criteria = this.currentSession().createCriteria(Sender.class)
				.add(Restrictions.like(FIELD_NAME, search + "%").ignoreCase());
		return list(criteria.addOrder(Order.asc(FIELD_NAME)));
	}

	public Optional<Sender> senderByName(String tagName) {
		Criteria criteria = this.currentSession().createCriteria(Sender.class)
				.add(Restrictions.eq(FIELD_NAME, tagName).ignoreCase());
		return Optional.ofNullable(uniqueResult(criteria));
	}

	public List<Sender> findAll() {
		Criteria criteria = this.currentSession().createCriteria(Sender.class);
		return list(criteria.addOrder(Order.asc(FIELD_NAME)));
	}
}

File: Api/src/main/kotlin/net/binggl/mydms/features/senders/SenderStore.kt
package net.binggl.mydms.features.senders

import org.springframework.jdbc.core.namedparam.MapSqlParameterSource
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate
import org.springframework.jdbc.support.GeneratedKeyHolder
import org.springframework.stereotype.Repository

@Repository
class SenderStore(private val jdbcT: NamedParameterJdbcTemplate) {

    fun findAll(): List<Sender> {
        return this.jdbcT.query("SELECT id,name FROM SENDERS", { rs, _ ->
            Sender(id = rs.getLong("id"), name = rs.getString("name"))
        })
    }

    fun search(name: String): List<Sender> {
        return this.jdbcT.query("SELECT id,name FROM SENDERS WHERE name LIKE :name",
                MapSqlParameterSource("name", "%$name%"), { rs, _ ->
            Sender(id = rs.getLong("id"), name = rs.getString("name"))
        })
    }

    fun save(tag: Sender): Sender {
        val holder = GeneratedKeyHolder()
        this.jdbcT.update("INSERT INTO SENDERS (name) VALUES(:name)",
                MapSqlParameterSource("name", tag.name),
                holder)
        return Sender(holder.key?.toLong(), tag.name)
    }
}